core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java

```diff
@@ -187,16 +191,12 @@ public MetaData applyChanges(MetaData oldMetaData, RoutingTable newRoutingTable)
                     .collect(Collectors.toSet());
             }
 
-            // only update in-sync allocation ids if there is at least one entry remaining. Assume for example that there only
-            // ever was a primary active and now it failed. If we were to remove the allocation id from the in-sync set, this would
-            // create an empty primary on the next allocation (see ShardRouting#allocatedPostIndexCreate)
-            if (inSyncAllocationIds.isEmpty() && oldInSyncAllocationIds.isEmpty() == false) {
-                assert updates.firstFailedPrimary != null :
-                    "in-sync set became empty but active primary wasn't failed: " + oldInSyncAllocationIds;
-                if (updates.firstFailedPrimary != null) {
-                    // add back allocation id of failed primary
-                    inSyncAllocationIds.add(updates.firstFailedPrimary.allocationId().getId());
-                }
+            // only remove allocation id of failed active primary if there is at least one active shard remaining. Assume for example that
+            // the primary fails but there is no new primary to fail over to. If we were to remove the allocation id of the primary from the
+            // in-sync set, this could create an empty primary on the next allocation.
+            if (newShardRoutingTable.activeShards().isEmpty() && updates.firstFailedPrimary != null) {
```

Comment:
If I understand you correctly, you mean "don't remove any allocation ids from the in-sync set if there are no active shards in the new routing table"? If so, that's not safe to do. For an explanation, see this test here that covers that scenario: https://github.com/elastic/elasticsearch/blob/e5f1eb5dc935695be088aaea6828fff66d250b4e/core/src/test/java/org/elasticsearch/cluster/routing/allocation/InSyncAllocationIdTests.java#L177-L183